### PR TITLE
setup docker

### LINKED ISSUE #2: setup Docker
On our first night getting this site setup, we ran into some installation and dependency issues. 

I'm currently stuck on this error, which seems like a problem with `libxml` and Homebrew:

`
Error: python-lxml has been disabled because it does not meet homebrew/core's requirements for Python library formulae! It was disabled on 2024-08-15.
`

Setting up Docker will likely resolve this and make setup for future users on local development much easier

### PR DESCRIPTION
**Pull request by derekeder**
Adds `Dockerfile` and `docker-compose.yml` for building and running this application locally.

Closes #2 

### Testing

You must have [Docker installed](https://docs.docker.com/engine/install/) on your machine.

- run `docker-compose build` and confirm the container builds
- after building, run the site locally with `docker-compose up`

### CODE REVIEW

**Comment by derekeder** [comment]
note: there seems to be an issue with running the app. likely something to do with the path to the working directory:


```bash
Attaching to chicago-elections-archive
chicago-elections-archive  |
chicago-elections-archive  | > chicago-elections-archive@0.1.0 start
chicago-elections-archive  | > NODE_ENV=development run-p -l start:*
chicago-elections-archive  |
chicago-elections-archive  | [start:rollup  ]
chicago-elections-archive  | [start:rollup  ] > chicago-elections-archive@0.1.0 start:rollup
chicago-elections-archive  | [start:rollup  ] > rollup --config rollup.config.mjs -w
chicago-elections-archive  | [start:rollup  ]
chicago-elections-archive  | [start:eleventy]
chicago-elections-archive  | [start:eleventy] > chicago-elections-archive@0.1.0 start:eleventy
chicago-elections-archive  | [start:eleventy] > eleventy --serve
chicago-elections-archive  | [start:eleventy]
chicago-elections-archive  | [start:postcss ]
chicago-elections-archive  | [start:postcss ] > chicago-elections-archive@0.1.0 start:postcss
chicago-elections-archive  | [start:postcss ] > postcss ./src/css -d ./site/_includes -w
chicago-elections-archive  | [start:postcss ]
chicago-elections-archive  | [start:rollup  ] rollup v3.14.0
chicago-elections-archive  | [start:rollup  ] bundles src/js/index.js → site/_includes/index.js...
chicago-elections-archive  | [start:rollup  ] babelHelpers: 'bundled' option was used by default. It is recommended to configure this option explicitly, read more here: https://github.com/rollup/plugins/tree/master/packages/babel#babelhelpers
chicago-elections-archive  | [start:eleventy] [11ty] Writing dist/style.json from ./site/style.11ty.js
chicago-elections-archive  | [start:eleventy] [11ty] Writing dist/manifest.json from ./site/manifest.json.njk
chicago-elections-archive  | [start:eleventy] [11ty] Writing dist/sitemap.xml from ./site/sitemap.xml.njk
chicago-elections-archive  | [start:eleventy] [11ty] Writing dist/robots.txt from ./site/robots.txt.njk
chicago-elections-archive  | [start:postcss ] Browserslist: caniuse-lite is outdated. Please run:
chicago-elections-archive  | [start:postcss ]   npx update-browserslist-db@latest
chicago-elections-archive  | [start:postcss ]   Why you should do it regularly: https://github.com/browserslist/update-db#readme
chicago-elections-archive  | [start:eleventy] [11ty] Problem writing Eleventy templates: (more in DEBUG output)
chicago-elections-archive  | [start:eleventy] [11ty] 1. Having trouble writing to "dist/404.html" from "./site/404.njk" (via EleventyTemplateError)
chicago-elections-archive  | [start:eleventy] [11ty] 2. (./site/_layouts/base.njk)
chicago-elections-archive  | [start:eleventy] [11ty]   Template render error: (/usr/src/app/site/_includes/assets.njk)
chicago-elections-archive  | [start:eleventy] [11ty]   EleventyShortcodeError: Error with Nunjucks shortcode `includefile` (via Template render error)
chicago-elections-archive  | [start:eleventy] [11ty] 3. ENOENT: no such file or directory, open '/usr/src/app/site/_includes/style.css' (via Template render error)
chicago-elections-archive  | [start:eleventy] [11ty]
chicago-elections-archive  | [start:eleventy] [11ty] Original error stack trace: Error: ENOENT: no such file or directory, open '/usr/src/app/site/_includes/style.css'
chicago-elections-archive  | [start:eleventy] [11ty]     at Object.openSync (node:fs:586:3)
chicago-elections-archive  | [start:eleventy] [11ty]     at Object.readFileSync (node:fs:454:35)
chicago-elections-archive  | [start:eleventy] [11ty]     at Object.<anonymous> (/usr/src/app/.eleventy.js:67:15)
chicago-elections-archive  | [start:eleventy] [11ty]     at Object.<anonymous> (/usr/src/app/node_modules/@11ty/eleventy/src/BenchmarkGroup.js:32:26)
chicago-elections-archive  | [start:eleventy] [11ty]     at ShortcodeFunction.run (/usr/src/app/node_modules/@11ty/eleventy/src/Engines/Nunjucks.js:228:35)
chicago-elections-archive  | [start:eleventy] [11ty]     at eval (eval at _compile (/usr/src/app/node_modules/nunjucks/src/environment.js:633:18), <anonymous>:27:71)
chicago-elections-archive  | [start:eleventy] [11ty]     at Template.root [as rootRenderFunc] (eval at _compile (/usr/src/app/node_modules/nunjucks/src/environment.js:633:18), <anonymous>:34:3)
chicago-elections-archive  | [start:eleventy] [11ty]     at Template.render (/usr/src/app/node_modules/nunjucks/src/environment.js:552:10)
chicago-elections-archive  | [start:eleventy] [11ty]     at eval (eval at _compile (/usr/src/app/node_modules/nunjucks/src/environment.js:633:18), <anonymous>:41:10)
chicago-elections-archive  | [start:eleventy] [11ty]     at fn (/usr/src/app/node_modules/a-sync-waterfall/index.js:26:24)
chicago-elections-archive  | [start:eleventy] [11ty] Copied 16 files / Wrote 4 files in 0.19 seconds (v2.0.0)
chicago-elections-archive  | ERROR: "start:eleventy" exited with 1.
```

**Comment by yashBhosale** [comment]
So I ran into a problem 
```
rollup v3.14.0
chicago-elections-archive  | bundles src/js/index.js → site/_includes/index.js...
chicago-elections-archive  | babelHelpers: 'bundled' option was used by default. It is recommended to configure this option explicitly, read more here: https://github.com/rollup/plugins/tree/master/packages/babel
#babelhelpers
chicago-elections-archive  | [!] (plugin commonjs--resolver) Error: EACCES: permission denied, stat '/root/.node_modules/src/js'
chicago-elections-archive  | Error: EACCES: permission denied, stat '/root/.node_modules/src/js'
chicago-elections-archive  |
chicago-elections-archive  |
chicago-elections-archive  |
chicago-elections-archive  | [2024-08-31 16:36:31] waiting for changes...
```
Which after several hours I was able to resolve by adding `RUN chown -R node /root` and then also a volume that mapped to `/root/.node_modules` in `docker-compose.yml`. I have basically 0 experience with setting up docker infra and/or navigating js build systems so there might be a more elegant way to handle that but that worked for me.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
+FROM node:17
```
Comment (yashBhosale) [inline Dockerfile:1]: It just so happens that node:17 ships with python 3.9.2 which is nice - if we ever feel the need to upgrade our python version then we might need to do some pyenv stuff.

**Review by yashBhosale — CHANGES_REQUESTED** [review]
Looks fine. my 2 review points are largely nitpicks but we do need to address the issue that you have. Also the fix for the problem I ran into (which I'm not sure why you didn't run into, because it seems like it would be universal based on https://github.com/rollup/plugins/discussions/1250)

**Comment by yashBhosale** [comment]
re the issue that you're having it seems like npm run start runs the eleventy, postcss, and rollup async rather than serially which feels like it could be causing issues?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+RUN poetry install
+RUN npm install
+RUN npm install -g mapshaper
+RUN git clone https://github.com/mapbox/tippecanoe.git
+RUN cd tippecanoe && make -j && make install
+
+# Bundle app source
+COPY . .
+
+RUN poetry run make all
```
[the remark this answers is not in the document]
Comment (derekeder) [reply Dockerfile:26]: good call - not needed

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+# A wildcard is used to ensure both package.json AND package-lock.json are copied
+# where available (npm@5+)
+COPY package*.json ./
+COPY pyproject.toml .
+
+RUN pip install poetry
+RUN pip install esridump
+RUN poetry install
+RUN npm install
+RUN npm install -g mapshaper
```
[the remark this answers is not in the document]
Comment (derekeder) [reply Dockerfile:19]: not sure why, but for some reason the makefile wouldn't build unless i put it here

**Comment by derekeder** [comment]
The original README is a bit light, but looking at `package.json` it seems like we want to run `npm run build && npm run start` to bring the site up. Unfortunately, there still seem to be some issues with it. 

@pjsier - could you share some details on how you typically run the site locally?

current stack trace:
```bash
derekeder@Rhodes chicago-elections-archive % docker-compose up
[+] Running 0/1
 ⠋ Container chicago-elections-archive  Recreated                                                                                            0.1s
Attaching to chicago-elections-archive
chicago-elections-archive  |
chicago-elections-archive  | > chicago-elections-archive@0.1.0 build
chicago-elections-archive  | > NODE_ENV=production run-s -l build:*
chicago-elections-archive  |
chicago-elections-archive  | [build:clean   ]
chicago-elections-archive  | [build:clean   ] > chicago-elections-archive@0.1.0 build:clean
chicago-elections-archive  | [build:clean   ] > rm -rf dist
chicago-elections-archive  | [build:clean   ]
chicago-elections-archive  | [build:rollup  ]
chicago-elections-archive  | [build:rollup  ] > chicago-elections-archive@0.1.0 build:rollup
chicago-elections-archive  | [build:rollup  ] > rollup --config rollup.config.mjs --bundleConfigAsCjs
chicago-elections-archive  | [build:rollup  ]
chicago-elections-archive  | [build:rollup  ]
chicago-elections-archive  | [build:rollup  ] src/js/index.js → site/_includes/index.js...
chicago-elections-archive  | [build:rollup  ] babelHelpers: 'bundled' option was used by default. It is recommended to configure this option explicitly, read more here: https://github.com/rollup/plugins/tree/master/packages/babel#babelhelpers
chicago-elections-archive  | [build:rollup  ] created site/_includes/index.js in 1.2s
chicago-elections-archive  | [build:postcss ]
chicago-elections-archive  | [build:postcss ] > chicago-elections-archive@0.1.0 build:postcss
chicago-elections-archive  | [build:postcss ] > postcss ./src/css -d ./site/_includes
chicago-elections-archive  | [build:postcss ]
chicago-elections-archive  | [build:postcss ] Browserslist: caniuse-lite is outdated. Please run:
chicago-elections-archive  | [build:postcss ]   npx update-browserslist-db@latest
chicago-elections-archive  | [build:postcss ]   Why you should do it regularly: https://github.com/browserslist/update-db#readme
chicago-elections-archive  | [build:postcss ]
chicago-elections-archive  | [build:postcss ] warn - The `purge`/`content` options have changed in Tailwind CSS v3.0.
chicago-elections-archive  | [build:postcss ] warn - Update your configuration file to eliminate this warning.
chicago-elections-archive  | [build:postcss ] warn - https://tailwindcss.com/docs/upgrade-guide#configure-content-sources
chicago-elections-archive  | [build:postcss ]
chicago-elections-archive  | [build:postcss ] 5:3	⚠  Replace fill-available to stretch, because spec had been changed [autoprefixer]
chicago-elections-archive  | [build:postcss ]
chicago-elections-archive  | [build:postcss ]
chicago-elections-archive  | [build:postcss ] 5:3	⚠  Replace fill-available to stretch, because spec had been changed [autoprefixer]
chicago-elections-archive  | [build:postcss ]
chicago-elections-archive  | [build:webpack ]
chicago-elections-archive  | [build:webpack ] > chicago-elections-archive@0.1.0 build:webpack
chicago-elections-archive  | [build:webpack ] > webpack --config webpack.config.js
chicago-elections-archive  | [build:webpack ]
chicago-elections-archive  | [build:webpack ] Browserslist: caniuse-lite is outdated. Please run:
chicago-elections-archive  | [build:webpack ]   npx update-browserslist-db@latest
chicago-elections-archive  | [build:webpack ]   Why you should do it regularly: https://github.com/browserslist/update-db#readme
chicago-elections-archive  | [build:webpack ] asset app.c716c7fca9c333993b14.js 186 KiB [emitted] [immutable] [minimized] (name: app) 2 related assets
chicago-elections-archive  | [build:webpack ] asset ../../site/_includes/webpack.njk 82 bytes [emitted]
chicago-elections-archive  | [build:webpack ] orphan modules 250 KiB [orphan] 27 modules
chicago-elections-archive  | [build:webpack ] runtime modules 221 bytes 1 module
chicago-elections-archive  | [build:webpack ] modules by path ./node_modules/core-js/internals/*.js 134 KiB
chicago-elections-archive  | [build:webpack ]   ./node_modules/core-js/internals/to-indexed-object.js 285 bytes [built] [code generated]
chicago-elections-archive  | [build:webpack ]   ./node_modules/core-js/internals/function-uncurry-this.js 365 bytes [built] [code generated]
chicago-elections-archive  | [build:webpack ]   + 174 modules
chicago-elections-archive  | [build:webpack ] modules by path ./node_modules/core-js/modules/*.js 125 KiB
chicago-elections-archive  | [build:webpack ]   ./node_modules/core-js/modules/es.array.iterator.js 2.56 KiB [built] [code generated]
chicago-elections-archive  | [build:webpack ]   ./node_modules/core-js/modules/es.object.to-string.js 397 bytes [built] [code generated]
chicago-elections-archive  | [build:webpack ]   + 55 modules
chicago-elections-archive  | [build:webpack ] modules by path ./src/js/*.js 249 KiB
chicago-elections-archive  | [build:webpack ]   ./src/js/legacy.js + 24 modules 249 KiB [built] [code generated]
chicago-elections-archive  | [build:webpack ]   ./src/js/embed.js 117 bytes [built] [code generated]
chicago-elections-archive  | [build:webpack ] ./node_modules/classlist-polyfill/src/index.js 5.56 KiB [built] [code generated]
chicago-elections-archive  | [build:webpack ] ./node_modules/url-search-params-polyfill/index.js 10.5 KiB [built] [code generated]
chicago-elections-archive  | [build:webpack ] webpack 5.75.0 compiled successfully in 3118 ms
chicago-elections-archive  | [build:eleventy]
chicago-elections-archive  | [build:eleventy] > chicago-elections-archive@0.1.0 build:eleventy
chicago-elections-archive  | [build:eleventy] > eleventy
chicago-elections-archive  | [build:eleventy]
chicago-elections-archive  | [build:eleventy] [11ty] Eleventy CLI Fatal Error: (more in DEBUG output)
chicago-elections-archive  | [build:eleventy] [11ty] 1. Error in your Eleventy config file '.eleventy.js'. (via EleventyConfigError)
chicago-elections-archive  | [build:eleventy] [11ty] 2. Cannot read properties of undefined (reading 'replace') (via TypeError)
chicago-elections-archive  | [build:eleventy] [11ty]
chicago-elections-archive  | [build:eleventy] [11ty] Original error stack trace: TypeError: Cannot read properties of undefined (reading 'replace')
chicago-elections-archive  | [build:eleventy] [11ty]     at Object.<anonymous> (/usr/src/app/site/_data/site.js:98:16)
chicago-elections-archive  | [build:eleventy] [11ty]     at Module._compile (node:internal/modules/cjs/loader:1099:14)
chicago-elections-archive  | [build:eleventy] [11ty]     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
chicago-elections-archive  | [build:eleventy] [11ty]     at Module.load (node:internal/modules/cjs/loader:975:32)
chicago-elections-archive  | [build:eleventy] [11ty]     at Function.Module._load (node:internal/modules/cjs/loader:822:12)
chicago-elections-archive  | [build:eleventy] [11ty]     at Module.require (node:internal/modules/cjs/loader:999:19)
chicago-elections-archive  | [build:eleventy] [11ty]     at require (node:internal/modules/cjs/helpers:102:18)
chicago-elections-archive  | [build:eleventy] [11ty]     at Object.<anonymous> (/usr/src/app/.eleventy.js:7:26)
chicago-elections-archive  | [build:eleventy] [11ty]     at Module._compile (node:internal/modules/cjs/loader:1099:14)
chicago-elections-archive  | [build:eleventy] [11ty]     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
chicago-elections-archive  | ERROR: "build:eleventy" exited with 1.
chicago-elections-archive  | npm notice
chicago-elections-archive  | npm notice New major version of npm available! 8.11.0 -> 10.8.3
chicago-elections-archive  | npm notice Changelog: https://github.com/npm/cli/releases/tag/v10.8.3
chicago-elections-archive  | npm notice Run npm install -g npm@10.8.3 to update!
chicago-elections-archive  | npm notice
chicago-elections-archive  |
chicago-elections-archive exited with code 1
```

**Comment by pjsier** [comment]
I would typically just use `npm start` without running `npm run build` because that's using `run-p` to start both `eleventy`, `rollup`, and `postcss`. It looks like what's failing here is that there are some environment variables that need to be set (and will be picked up by a `.env` file).

The one it's failing on here is `SITE_HOST` because `npm run build` runs as `NODE_ENV=production` which expects that to be present. You could default it here or set that locally and it should pass. I generally only used `build` for deploying and you shouldn't need it for general local development

https://github.com/chihacknight/chicago-elections-archive/blob/b89e7d0d50acd50b721dd4eb1463266516915e89/site/_data/site.js#L5

**Comment by derekeder** [comment]
@pjsier excellent - thanks for the tips! Setting the `.env` file and variables did the trick and the site now comes up. I can get the `/about/` and `/download/` pages, but the root `/en/` page shows a 404 not found.

Screenshot 2024-09-09 at 3 18 17 PM

I see that `AZURE_MAPS_KEY` is also listed in `.env.sample`. The site looks to use `maplibre` - what role does Azure play? Do we need to set up an account and set that key?

**Comment by pjsier** [comment]
I don't think `/en/` was ever set up? I looked into it but I don't believe it was ever finished and the currently published site is sitting at the root `/`

Azure maps is just used for the geocoder, so if you don't want to use it you should be able to just delete that component

**Comment by derekeder** [comment]
@pjsier thx! you're right lol - I had that route cached from another site I was working on that was also using port 8080. all looks good now!

And thanks for the info on Azure. We'll see if we want to keep it or use something else.

**Comment by derekeder** [comment]
@yashBhosale this is ready for another look. let me know if it works for you!

**Pull request merged**